cpp/include/raft/neighbors/cagra_types.hpp

```diff
@@ -183,15 +185,36 @@ struct index : ann::index {
   {
     RAFT_EXPECTS(dataset.extent(0) == knn_graph.extent(0),
                  "Dataset and knn_graph must have equal number of rows");
+    RAFT_LOG_INFO("CAGRA constructor, copying arrays");
     raft::copy(dataset_.data_handle(), dataset.data_handle(), dataset.size(), res.get_stream());
     raft::copy(graph_.data_handle(), knn_graph.data_handle(), knn_graph.size(), res.get_stream());
+    dataset_view_ = dataset_.view();
+    graph_view_   = graph_.view();
     res.sync_stream();
   }
 
+  /** Construct an index from dataset and knn_graph arrays */
+  index(raft::device_resources const& res,
+        raft::distance::DistanceType metric,
+        raft::device_matrix_view<const T, IdxT, row_major> dataset,
+        raft::device_matrix_view<const IdxT, IdxT, row_major> knn_graph)
```

Comment:
Right now, passing `device_matrix_views` to the constructor would select this specialization (assuming that `const` qualifiers are not forgotten). Shall we make it more explicit for the user to select which constructor to call (e.g. by an additional param / template param)?

Reply:
I think we should just add more explicit docs

Reply:
In the context of this PR extending the docs sounds good, but in #1505, even `device_mdspan` input might be copied if the number of dimensions would not be compatible with vectorized loads. Do you think adding clear docs is still the best option? https://github.com/rapidsai/raft/pull/1505/files#diff-0512c89e79e0d6300fda20bf29340258cd48ce7c5b5b81f4fbad61e8eae3db5aR191-R206

Reply:
I do think so, yes, with a tiny code example in the associated docs. For the first constructor, perhaps we can use host_mdspan instead of vanilla mdspan? Then the overloads will be automatically selected based on the input, and if there's any instances where the code queries the data using driver/rt APIs to check where a pointer is located, we can remove that entirely. What are your thoughts?

Reply:
We should probably explicitly add the constructors to the docs source that we intend for users to interact with.

Small thing I noticed while looking over this PR: device_resources in the constructor should be changed to raft::resources.